packages/core/src/factories/apiClientFactory/index.ts

```diff
@@ -23,7 +23,7 @@ const apiClientFactory = <ALL_SETTINGS extends ApiClientConfig, ALL_FUNCTIONS>(f
       .filter(ext => isFn(ext.beforeCreate))
       .reduce((prev, curr) => curr.beforeCreate({ configuration: prev }), config);
 
-    const settings = factoryParams.onCreate ? factoryParams.onCreate(_config) : { config, client: config.client };
+    const settings = factoryParams.onCreate ? await factoryParams.onCreate(_config) : { config, client: config.client };
```

Comment:
Here is the change I suggested during our daily, exposing the `req` and `res` properties on `onCreate`:

```suggestion
    const settings = factoryParams.onCreate ? await factoryParams.onCreate(_config, this.middleware) : { config, client: config.client };
```

In general, according to the `ApiClientFactoryParams` type, the `onCreate` function should take 2 params:

```
export interface ApiClientFactoryParams<T, F = any> {
  ...
  onCreate: (config: T, headers?: Record<string, string>) => Promise<{ config: T; client: any }>;
  ...
}
```

Without the above change, `headers` are `undefined` and for the purpose of multistore functionality, we need the `req` to pick up the origin hostname. In the **dev** environment, we also need `res`, as there are problems with CORS, and we need a proper `Access-Control-Allow-Origin` header dynamically generated and appended to the response based on the hostname where the request came from.

Reply:
From what I understand, we need an optional req and/or res object. 
Can't we use already existing hooks like `beforeCreate` or `hook` function to set/get whatever you need?
The fact that we do not pass headers there is one thing, but passing the whole middleware will make everything even vaguer. I really do not want to do that. Middleware is a singleton for that process and so are the req and res objects on the middleware, both available in hooks. Please, consider such a solution.